tools/doc-gen/index.ts

```diff
@@ -0,0 +1,377 @@
+          ) {
+            caseInfo['code'] = codeNode.initializer.text;
+          } else {
+            // No code, so abort
+            return;
+          }
+
+          invalidCases.push(caseInfo);
+        }
+      });
+    }
+    ts.forEachChild(node, findTestCases);
+  };
+
+  findTestCases(specSource);
+
+  // Get rule file exports
+  const { rule, RULE_NAME } = await import(ruleInfo.rule);
+
+  const ruleConfig = rule as TSESLint.RuleModule<string, []> & {
+    defaultOptions?: Record<string, unknown>[];
+  };
+
+  return {
+    name: RULE_NAME,
+    files: ruleInfo,
+    description: ruleConfig.meta.docs?.description,
+    type: ruleConfig.meta.type,
+    validCases,
+    invalidCases,
+  };
+}
+
+/**
+ * Generate markdown documentation for given rule
+ * (Goes to path stored in outputPath)
+ * @param name rule display name
+ * @param info file info, including base file name
+ * @param validCases
+ * @param invalidCases
+ */
+function generateDocumentation(ruleData: RuleData) {
+  const valid = ruleData.validCases.map((testCase) => {
+    return `### ${testCase.name}
+
+\`\`\`ts
+${normalizeCodeBlockIndentation(testCase.code)}
+\`\`\`
+
+`;
+  }).join(`
+`);
+
+  const invalid = ruleData.invalidCases.map((testCase) => {
+    return `### ${testCase.name}
+
+\`\`\`ts
+${normalizeCodeBlockIndentation(testCase.code)}
+\`\`\`
+
+`;
+  }).join(`
+`);
+
+  const md = `# Rule \`${ruleData.name}\`
+
+${normalizeMarkdownText(ruleData.description ?? '')}
+
+## Valid Usage
+
+${valid ? valid : 'No test cases'}
+
+## Invalid Usage
+
+${invalid ? invalid : 'No test cases'}
+`;
+
+  if (!existsSync(outputPath)) {
+    mkdirSync(outputPath, { recursive: true });
+  }
+  writeFileSync(join(outputPath, `${ruleData.files.baseName}.md`), md);
+}
+
+async function generateDocsForRules(ruleFileInfo: RuleFileInfo[]) {
+  for (const rule of ruleFileInfo) {
+    const data = await readRuleData(rule);
+    generateDocumentation(data);
+  }
+}
+
+/**
+ * Remove leading indentation on code block while maintaining overall
+ * indentation. Also cleans up leading or trailing blank line
+ * @param codeblock Code block to clean up
+ * @returns
+ */
+function normalizeCodeBlockIndentation(codeblock: string): string {
+  const indentSize = (line: string) => (line.match(/^[\s]+/) ?? [''])[0].length;
+  const lines = codeblock
+    .replace(/^\n/, '')
+    .replace(/\n[\s]*$/, '')
+    .split('\n');
+
+  const minIndent = Math.min(...lines.filter((l) => l.length).map(indentSize));
+  return lines.map((l) => l.substring(minIndent)).join('\n');
+}
+
+/**
+ * Escapes special markdown characters and normalizes indentation on non-markdown
+ * text
+ * @param text
+ * @returns
+ */
+function normalizeMarkdownText(text: string): string {
+  return text
+    .replace(/\*/g, '\\*')
+    .replace(/#/g, '\\#')
+    .split('\n')
+    .map((l) => l.trim())
+    .join('\n');
+}
+
+// -----
+
+const rulesInfo = gatherRuleFileInfo(
+  join(__dirname, '../eslint-rules/rules/angular')
```

Comment:
not a blocker since I think we all use macs anyway, but is there a more multi platform way to get __dirname? I'm not familiar to how people do this on windows without bash.

Reply:
I assumed `__dirname` _is_ multiplatform. I believe it's a node thing, and thus how it resolves would depend on your platform

Reply:
path and dirname are node things, but __dirname isn't: https://stackoverflow.com/a/28279609

Reply:
I think that's only in ES modules. The proposed workaround doesn't actually work when I try adding it to the code as it is currently configured:

```
TSError: ⨯ Unable to compile TypeScript:
tools/doc-gen/index.ts:16:34 - error TS1343: The 'import.meta' meta-property is only allowed when the '--module' option is 'es2020', 'es2022', 'esnext', 'system', 'node16', or 'nodenext'.

16 const __filename = fileURLToPath(import.meta.url);
```

Reply:
(the tsconfig in the tool directory is set to use commonjs modules and target es5 currently)

Reply:
I think it'll become a moot point once we move to an executor in any case